doc/RawSyscallsDesign.md

```diff
@@ -0,0 +1,196 @@
+   compiler it doesn't exit, and it should be obvious the return values are all
+   unused.
+1. The compiler is unable to optimize away unused arguments. For example,
+   Memop's "get process RAM start address" operation only needs `r0` set, while
+   Memop's "break" operation needs both `r0` and `r1` set. If our inline
+   assembly calls "get process RAM start address" but sets both `r0` and `r1`,
+   the compiler doesn't know that `r1` will be ignored so setting that register
+   will not be optimized away. Therefore we want to set the minimum number of
+   argument registers possible.
+1. The cost of specifying unused return registers is only that of unnecessarily
+   marking a register as clobbered. Explanation: After inlining, an unused
+   register is marked as "changed by the assembly" but can immediately be
+   re-used by the compiler, which is the same as a clobbered register. System
+   calls should generally be inlined -- and even if they aren't, the unused
+   return values will probably be passed in caller-saved registers (this is true
+   for the C ABI, so probably true for the Rust ABI), which are treated as
+   clobbered regardless.
+1. Command always needs to set all four registers, because otherwise we risk
+   leaking data to capsules (which are untrusted for confidentiality). Exit,
+   Memop, and Yield do not have this concern because they are implemented by the
+   core kernel (which is trusted).
+
+## Yield Method Breakdown
+
+`yield-wait` does not take any arguments, while `yield-no-wait` takes 1.
+Therefore, to avoid setting arguments unnecessarily, `yield-wait` and
+`yield-no-wait` should be distinct functions.
+
+## Remaining Method Breakdown
+
+After taking care of `yield`, we break down the remaining system calls by the
+number of arguments they take. Here is the breakdown (Exit and Memop are listed
+twice because they have sub-operations with varying number of arguments).
+
+* **Zero:**
+* **One:** Exit, Memop
+* **Two:** Exit, Memop
+* **Three:**
+* **Four:** Command, Read-Only Allow, Read-Write Allow, Subscribe
+
+Therefore we need three more methods: `one_arg_syscall`, `two_arg_syscall`,
+`four_arg_syscall`.
+
+## `u32` versus `usize`
+
+The decision of where to use `u32` and where to use `usize` can be a bit tricky.
+The Tock syscall ABI is currently only specified for 32-bit systems, so on real
+Tock systems both types match the size of a register, but the unit test
+environment can be either 32 bit or 64 bit. `RawSyscalls` uses `usize` for
+values that can contain pointers, so that pointers are not truncated in the unit
+test environment. To keep types as consistent as possible, it uses `u32` for
+register-sized values that cannot be pointers.
+
+## Yield Methods Design
+
+Rust does not specify the representation of `bool`, only that it casts to `0`
+and `1`, so it would not be correct use a `bool` as the flag for
+`yield-no-wait`. Instead we use a `u8`.
```

Comment:
Rust *does* (or plans to) specify it to be equivalent to `_Bool` in C17, which among other things means that the expected values are actually correct. UGW has some text on this: https://rust-lang.github.io/unsafe-code-guidelines/layout/scalars.html#bool.

In particular, see the true/false macros in https://en.cppreference.com/w/c/types.

(Though I don't really care which one of these you prefer.)

Reply:
Hmm, I don't see anything saying `bool` would have that representation. If you interpret the document you link as making a claim about the representation of `bool`, then I dispute your interpretation.

Regardless, I left it at `u8`. I could use an enum like `#[repr(u8)] enum Flag { No = 0, Yes = 1 }` but that adds an additional form of unsafety to `raw_yield_no_wait`.